rules/prefer-global-this.js

```diff
@@ -0,0 +1,491 @@
+	'originAgentCluster',
+	'alert',
+	'confirm',
+	'prompt',
+	'print',
+	'postMessage',
+
+	// global events
+	'onabort',
+	'onauxclick',
+	'onbeforeinput',
+	'onbeforematch',
+	'onbeforetoggle',
+	'onblur',
+	'oncancel',
+	'oncanplay',
+	'oncanplaythrough',
+	'onchange',
+	'onclick',
+	'onclose',
+	'oncontextlost',
+	'oncontextmenu',
+	'oncontextrestored',
+	'oncopy',
+	'oncuechange',
+	'oncut',
+	'ondblclick',
+	'ondrag',
+	'ondragend',
+	'ondragenter',
+	'ondragleave',
+	'ondragover',
+	'ondragstart',
+	'ondrop',
+	'ondurationchange',
+	'onemptied',
+	'onended',
+	'onerror',
+	'onfocus',
+	'onformdata',
+	'oninput',
+	'oninvalid',
+	'onkeydown',
+	'onkeypress',
+	'onkeyup',
+	'onload',
+	'onloadeddata',
+	'onloadedmetadata',
+	'onloadstart',
+	'onmousedown',
+	'onmouseenter',
+	'onmouseleave',
+	'onmousemove',
+	'onmouseout',
+	'onmouseover',
+	'onmouseup',
+	'onpaste',
+	'onpause',
+	'onplay',
+	'onplaying',
+	'onprogress',
+	'onratechange',
+	'onreset',
+	'onresize',
+	'onscroll',
+	'onscrollend',
+	'onsecuritypolicyviolation',
+	'onseeked',
+	'onseeking',
+	'onselect',
+	'onslotchange',
+	'onstalled',
+	'onsubmit',
+	'onsuspend',
+	'ontimeupdate',
+	'ontoggle',
+	'onvolumechange',
+	'onwaiting',
+	'onwebkitanimationend',
+	'onwebkitanimationiteration',
+	'onwebkitanimationstart',
+	'onwebkittransitionend',
+	'onwheel',
+
+	// Window events
+	'onafterprint',
+	'onbeforeprint',
+	'onbeforeunload',
+	'onhashchange',
+	'onlanguagechange',
+	'onmessage',
+	'onmessageerror',
+	'onoffline',
+	'ononline',
+	'onpagehide',
+	'onpagereveal',
+	'onpageshow',
+	'onpageswap',
+	'onpopstate',
+	'onrejectionhandled',
+	'onstorage',
+	'onunhandledrejection',
+	'onunload',
+
+	// https://www.w3.org/TR/DOM-Level-2-Events/events.html#Events-flow
+	'addEventListener',
+	'removeEventListener',
+	'dispatchEvent',
+
+	// https://dom.spec.whatwg.org/#idl-index
+	'Event',
+	'event',
+	'CustomEvent',
+	'EventTarget',
+	'AbortController',
+	'AbortSignal',
+	'NodeList',
+	'HTMLCollection',
+	'MutationObserver',
+	'MutationRecord',
+	'EventTarget',
+	'Document',
+	'XMLDocument',
+	'DOMImplementation',
+	'DocumentType',
+	'DocumentFragment',
+	'ShadowRoot',
+	'Element',
+	'NamedNodeMap',
+	'Attr',
+	'CharacterData',
+	'Text',
+	'CDATASection',
+	'ProcessingInstruction',
+	'Comment',
+	'AbstractRange',
+	'StaticRange',
+	'Range',
+	'NodeIterator',
+	'TreeWalker',
+	'NodeFilter',
+	'DOMTokenList',
+	'XPathResult',
+	'XPathExpression',
+	'XPathEvaluator',
+	'XSLTProcessor',
+
+	// https://drafts.csswg.org/cssom-view/#idl-index
+	'matchMedia',
+	'screen',
+	'visualViewport',
+	'moveTo',
+	'moveBy',
+	'resizeTo',
+	'resizeBy',
+	'innerWidth',
+	'innerHeight',
+	'scrollX',
+	'pageXOffset',
+	'scrollY',
+	'pageYOffset',
+	'scroll',
+	'scrollTo',
+	'scrollBy',
+	'screenX',
+	'screenLeft',
+	'screenY',
+	'screenTop',
+	'screenWidth',
+	'screenHeight',
+	'devicePixelRatio',
+	'MediaQueryList',
+	'MediaQueryListEvent',
+	'Screen',
+	'CaretPosition',
+	'VisualViewport',
```

Comment:
Note that this list is somewhat arbitrary. My logic for things that belong here:

- it's directly related to window (✅ `window.close()`)
- it does NOT work well as `globalThis.x` or `x` (✅ `window.frames`, `window.top`)

More easily, I removed:

- all events that bubble up (❌ `window.onplay`)
- all events that works just as well on a "global" (✅ `globalThis.online`)
- all constructors

Note that some constructors are occasionally related to window (like Element) but works just as well as globals, unless you're accessing `iframe.contentWindow.Element`, which is not touched.

```suggestion
	// Properties and methods
	// https://html.spec.whatwg.org/multipage/nav-history-apis.html#the-window-object
	'name',
	'locationbar',
	'menubar',
	'personalbar',
	'scrollbars',
	'statusbar',
	'toolbar',
	'status',
	'close',
	'closed',
	'stop',
	'focus',
	'blur',
	'frames',
	'length',
	'top',
	'opener',
	'parent',
	'frameElement',
	'open',
	'originAgentCluster',
	'postMessage',

	// Events commonly associated with "window"
	'onresize',
	'onblur',
	'onfocus',
	'onload',
	'onscroll',
	'onscrollend',
	'onwheel',
	'onbeforeunload', // Browsers might have specific behaviors on exactly `window.onbeforeunload =`
	'onmessage',
	'onmessageerror',
	'onpagehide',
	'onpagereveal',
	'onpageshow',
	'onpageswap',
	'onunload',

	// To add/remove/dispatch events that are commonly associated with "window"
	// https://www.w3.org/TR/DOM-Level-2-Events/events.html#Events-flow
	'addEventListener',
	'removeEventListener',
	'dispatchEvent',

	// https://dom.spec.whatwg.org/#idl-index
	'event', // Deprecated and quirky, best left untouched

	// https://drafts.csswg.org/cssom-view/#idl-index
	'screen',
	'visualViewport',
	'moveTo',
	'moveBy',
	'resizeTo',
	'resizeBy',
	'innerWidth',
	'innerHeight',
	'scrollX',
	'pageXOffset',
	'scrollY',
	'pageYOffset',
	'scroll',
	'scrollTo',
	'scrollBy',
	'screenX',
	'screenLeft',
	'screenY',
	'screenTop',
	'screenWidth',
	'screenHeight',
	'devicePixelRatio',
```

Reply:
Can you turn this message into a comment so we know why some properties are missing and when some should not be added? Because I expect this list to receive some updates and we want to make sure that they make sense.

Reply:
Add to comment a496c6f3ab6fadf7dae952a4bdaa4227e6cf2068